### PR TITLE
IDAM-106 Redis

### PR DESCRIPTION
**Pull request by niceHPAmido**
Reconfigured the .net core sdk to implement a startup base class for each client
Reconfigured the .net core test client to consume SDK startup base class
Added Redis cache and local configuration to SDK

### CODE REVIEW

File: NICE.Identity.Authentication.Sdk/NICE.Identity.Authentication.Sdk.csproj [hunk 1]
```diff
@@ -12,15 +12,17 @@
...
+    <PackageReference Include="Microsoft.AspNetCore.Authentication.Cookies" Version="2.2.0" />
+    <PackageReference Include="Microsoft.AspNetCore.Authentication.OpenIdConnect" Version="2.2.0" />
+    <PackageReference Include="Microsoft.AspNetCore.DataProtection.StackExchangeRedis" Version="2.2.0" />
+    <PackageReference Include="Microsoft.AspNetCore.Mvc" Version="2.2.0" />
+    <PackageReference Include="Microsoft.AspNetCore.Session" Version="2.2.0" />
+    <PackageReference Include="Microsoft.Extensions.Configuration" Version="2.2.0" />
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="2.2.0" />
+    <PackageReference Include="Microsoft.Extensions.Http" Version="2.2.0" />
+    <PackageReference Include="Microsoft.Extensions.Options" Version="2.2.0" />
+    <PackageReference Include="Microsoft.Extensions.Options.ConfigurationExtensions" Version="2.2.0" />
```
Comment (John-Williams) [inline NICE.Identity.Authentication.Sdk/NICE.Identity.Authentication.Sdk.csproj]: looks like Microsoft.AspNetCore.Authorization below this line is still set to 2.1

File: NICE.Identity.Test/NICE.Identity.Test.csproj [hunk 2]
```diff
@@ -1,24 +1,24 @@
...
+    <PackageReference Include="Microsoft.AspNetCore.Mvc.Abstractions" Version="2.2.0" />
+    <PackageReference Include="Microsoft.AspNetCore.Mvc.Testing" Version="2.2.0" />
+    <PackageReference Include="Microsoft.AspNetCore.TestHost" Version="2.2.0" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.InMemory" Version="2.2.2" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="2.2.2" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.UserSecrets" Version="2.2.0" />
+    <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="2.2.0" />
+    <PackageReference Include="Microsoft.Extensions.Logging.Debug" Version="2.2.0" />
+    <PackageReference Include="Microsoft.Extensions.Options.ConfigurationExtensions" Version="2.2.0" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.0.0" />
```
Comment (John-Williams) [inline NICE.Identity.Test/NICE.Identity.Test.csproj]: there's still other 2.1 references in this file outside of these changes. can they all be moved please?

File: TestClients/NICE.Identity.TestClient.NETCore/Program.cs [hunk 3]
```diff
@@ -1,17 +1,36 @@
...
-		}
+			Func<IHostingEnvironment, IConfigurationBuilder> configurationFactory = env =>
+				new ConfigurationBuilder()
+					.SetBasePath(env.ContentRootPath)
+					.AddEnvironmentVariables()
+					.AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
+					//.AddJsonFile($"appsettings.{environmentName}.json", optional: true)
+					.AddUserSecrets(Assembly.GetAssembly(typeof(Startup)));
+
+            var startup = new Startup("Publications", configurationFactory, ProductionDependencies.AddProductionDependencies);
```
Comment (John-Williams) [inline TestClients/NICE.Identity.TestClient.NETCore/Program.cs]: i'm not sure about this line. why is this needed?

File: TestClients/NICE.Identity.TestClient.NETCore/Program.cs [hunk 4]
```diff
@@ -1,17 +1,36 @@
...
 namespace NICE.Identity.TestClient.NETCore
 {
 	public class Program
 	{
 		public static void Main(string[] args)
 		{
-			CreateWebHostBuilder(args).Build().Run();
-		}
+			Func<IHostingEnvironment, IConfigurationBuilder> configurationFactory = env =>
+				new ConfigurationBuilder()
```
Comment (John-Williams) [inline TestClients/NICE.Identity.TestClient.NETCore/Program.cs]: why not use CreateDefaultBuilder?

File: NICE.Identity.Authentication.Sdk/Extensions/ServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -1,21 +1,50 @@
...
+using StackExchange.Redis;
 using IConfiguration = Microsoft.Extensions.Configuration.IConfiguration;
 
-namespace NICE.Identity.Authentication.Sdk
+namespace NICE.Identity.Authentication.Sdk.Extensions
 {
 	public static class ServiceCollectionExtensions
     {
-        public static IServiceCollection AddAuthenticationSdk(this IServiceCollection services,
+	    public static IServiceCollection AddRedisCacheSDK(this IServiceCollection services,
```
Comment (John-Williams) [inline NICE.Identity.Authentication.Sdk/Extensions/ServiceCollectionExtensions.cs]: afaik IServiceCollection isn't available on .net framework. 
should this code also go in AppBuilder extensions? and if so, could it be refactored so it's not duplicated?

File: NICE.Identity.Authentication.Sdk/Extensions/ServiceCollectionExtensions.cs [hunk 6]
```diff
@@ -1,21 +1,50 @@
...
+            services.AddDataProtection()
+                    .SetApplicationName(clientName)
+                    .PersistKeysToStackExchangeRedis(redis, $"{clientName}.DataProtection-Keys");
+
+            services.AddSession(options =>
+		    {
+			    options.Cookie.SecurePolicy = CookieSecurePolicy.Always;
+			    options.Cookie.Name = $"{clientName}.Session";
+			    options.Cookie.HttpOnly = true;
+			    options.IdleTimeout = TimeSpan.FromMinutes(10);
```
[the remark this answers is not in the document]
Comment (niceHPAmido) [reply NICE.Identity.Authentication.Sdk/Extensions/ServiceCollectionExtensions.cs]: The IdleTimeout indicates how long the session can be idle before its contents are abandoned. Each session access resets the timeout. Note this only applies to the content of the session, not the cookie.

File: NICE.Identity.Authentication.Sdk/Extensions/ServiceCollectionExtensions.cs (hunk 5 above)
Comment (niceHPAmido) [reply NICE.Identity.Authentication.Sdk/Extensions/ServiceCollectionExtensions.cs]: Agreed

File: TestClients/NICE.Identity.TestClient.NETCore/Program.cs (hunk 3 above)
Comment (niceHPAmido) [reply TestClients/NICE.Identity.TestClient.NETCore/Program.cs]: You need this line to add production dependencies but also swap out the dependencies for mocked objects for testing.

File: TestClients/NICE.Identity.TestClient.NETCore/Program.cs (hunk 4 above)
Comment (niceHPAmido) [reply TestClients/NICE.Identity.TestClient.NETCore/Program.cs]: Under the hood the same calls are being made. We have more control of the behavior by making the call this way.

File: NICE.Identity.Authentication.Sdk/Extensions/ServiceCollectionExtensions.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (John-Williams) [reply NICE.Identity.Authentication.Sdk/Extensions/ServiceCollectionExtensions.cs]: can the timeout be a configured value please?
